### ISSUE TITLE
A component required a bean named 'kubernetesConfigRetryInterceptor' that could not be found

### ISSUE THREAD

**Issue by vxavictor513** [spring-cloud/spring-cloud-kubernetes#963]
Hi, I'm upgrading Spring Cloud from 2020.0.1 to 2021.0.0 and hit the following error when starting up the application.

Any idea what could have gone wrong?

```
***************************
APPLICATION FAILED TO START
***************************

Description:

A component required a bean named 'kubernetesConfigRetryInterceptor' that could not be found.


Action:

Consider defining a bean named 'kubernetesConfigRetryInterceptor' in your configuration.
```

I am using `spring-cloud-starter-kubernetes-fabric8-all` dependency and below is the sample configuration I have in `bootstrap.properties` (yes, it is using `spring-cloud-starter-bootstrap`).

```
spring.cloud.kubernetes.enabled=true
spring.cloud.kubernetes.discovery.enabled=true
spring.cloud.kubernetes.config.enabled=true
spring.cloud.kubernetes.config.sources[0].name=configmap1
spring.cloud.kubernetes.config.sources[0].namespace=dev
spring.cloud.kubernetes.config.sources[1].name=configmap2
spring.cloud.kubernetes.config.sources[1].namespace=dev

spring.cloud.kubernetes.reload.enabled=true
spring.cloud.kubernetes.reload.strategy=restart_context
spring.cloud.kubernetes.reload.monitoring-config-maps=false
spring.cloud.kubernetes.reload.mode=event
```

**Comment by wind57** [comment]
> ... hit the following error when starting up the application

locally, I assume? Not inside kubernetes?

**Comment by vxavictor513** [comment]
> > ... hit the following error when starting up the application
> 
> locally, I assume? Not inside kubernetes?

@wind57: Sorry for not stating clearly, it was running on Kubernetes.

And I just realize that if I set `spring.cloud.kubernetes.config.fail-fast=true` then this error is gone.

**Comment by wind57** [comment]
any updates on this? this retry logic has been implemented recently, and it seems you are the first reporter of a problem. It would help us a lot if you could provide a way to reproduce this. thank you upfront

**Comment by vxavictor513** [comment]
Sorry, I'm trying to reproduce it in a local Kubernetes but I couldn't. It might take some time.

**Comment by vxavictor513** [comment]
@wind57: Here you go, https://github.com/vxavictor513/spring-cloud-kubernetes-issue-963-demo

Hope this helps. Let me know if anything else I can assist. 😃

**Comment by wind57** [comment]
massive thank you for the sample! It is exactly what I had a hint for, initially.

There are a series of events that you got stuck into, I'll try to explain. 

So, the way we implement retry is via `spring-retry`, which in turn (simplified) is enabled via two annotations: `@Retryable` and `@EnableRetry`. In our kubernetes project we have a method (simplified again) like this:

```
@Retryable(interceptor="kubernetesConfigRetryInterceptor")
public PropertySource<?> locate(....) {....}
```

and there is a configuration that will provide `@EnableRetry` (I've only showed the relevant parts here):

```
@ConditionalOnKubernetesConfigOrSecretsRetryEnabled
@EnableRetry(proxyTargetClass = true)
static class RetryConfiguration {
      		@Bean("kubernetesConfigRetryInterceptor")
		public RetryOperationsInterceptor kubernetesConfigRetryInterceptorNoRetry() {
			return .....
		}
}
```

In plain english we say : you want retry enabled? Sure, please provide some flags  (for the `@ConditionalOnKubernetesConfigOrSecretsRetryEnabled` annotation that is). 

So we are "protected": unless users request retry on demand (documentation has a lot more details) - we are good.

---

But, you bring `consul` into the dependencies :) and it messes this logic via its [ConsulAutoConfiguration.RetryConfiguration](https://github.com/spring-cloud/spring-cloud-consul/blob/main/spring-cloud-consul-core/src/main/java/org/springframework/cloud/consul/ConsulAutoConfiguration.java#L113). Notice that it also has `@ConditionalOnClass({ Retryable.class, Aspect.class, AopAutoConfiguration.class })` and these dependencies are brought in transitively by `spring-boot-starter-batch` and `spring-boot-starter-aop`. Kind of crazy how the stars have aligned here, I admit.  

Because it has this property : `@ConditionalOnProperty(value = "spring.cloud.consul.retry.enabled", matchIfMissing = true)`, `@EnableRetry` happens, _without_ our `@ConditionalOnKubernetesConfigOrSecretsRetryEnabled` being enable.

So we reach `@Retryable(interceptor="kubernetesConfigRetryInterceptor")` (remember that `@EnableRetry` has kicked in also), we need a `kubernetesConfigRetryInterceptor` bean, but because `@ConditionalOnKubernetesConfigOrSecretsRetryEnabled` is not active, the config that is supposed to provide that bean is not active either and voilà! Exception when starting the context.

---

The fact that you set `spring.cloud.kubernetes.config.fail-fast=true` means that you activate our config... and it works; though I do not quite agree that this is the correct solution. The correct solution is : "it depends" :) 

 - If you do _not_ want retry on the consul side, put this `spring.cloud.consul.retry.enabled=false` in `bootstrap.properties` of your project. Though, I am kind of inclined to say that this is a bug in consul - do they really want to provide retry enabled _by default_? I would argue that this : `@ConditionalOnProperty(value = "spring.cloud.consul.retry.enabled", matchIfMissing = true)` should really be `@ConditionalOnProperty(value = "spring.cloud.consul.retry.enabled")` may be? Hard to say...

- if you do want consul to retry, but you do not want spring kubernetes to retry, you should be adding _two_ properties:
`spring.cloud.kubernetes.config.fail-fast=true` and `spring.cloud.kubernetes.config.retry.enabled=false`.

- if you want retry in both consul and kubernetes: `spring.cloud.kubernetes.config.fail-fast=true` and `spring.cloud.kubernetes.config.retry.enabled=true` - but this one is implicit. 

imho, your issue males me question the names we have for retry. I'll think this over a little and may be raise a defect.

@ryanjbaxter in the meanwhile - wdyt about `consul` config with auto-enabled retry?

**Comment by wind57** [comment]
There is a problem here though that I have realized over the weekend. Again, it will take a while to explain :( 

Suppose, we strip out all your dependencies and leave only the kubernetes fabric8 one. You also add a configmap to the source to be read, that simply does not exist. In this case, without any additional settings, your application will still start. Yes, you will get a warning in the logs saying that a certain configmap could not be read, nevertheless, your app would start just fine. This is the logic in place in our project for a very long time.

Now, add consul into the picture. You are _forced_ to set `spring.cloud.kubernetes.config.fail-fast=true` and it does not matter if you now set `spring.cloud.kubernetes.config.retry.enabled=true/false` - your application will not start; because we will fail it.

This is kind of a big inconsistency, imho. I'll spent some time thinking how to solve this, if possible at all.

**Comment by wind57** [comment]
I am proposing [some changes for this issue](https://github.com/spring-cloud/spring-cloud-kubernetes/pull/964), because to me this is a valid defect. 

Basically, I am inclined to say : 

 - we _always_ enable spring-retry (unlike what we did so far)
 - by default, all the retry interceptors are going to have `NeverRetryPolicy`, thus we are consistent with what happened before.
 - if you want explicit retry for secrets/config, you can enable it.
 - `fail-fast` works the same way.

@vxavictor513 imho: this is a valid defect. We need to wait for input here, though.

**Comment by ryanjbaxter** [comment]
Would it be simpler and cleaner if we just had a Retryable implementation of `KuberenetesClientConfigMaPropertySounceLocator` and `Fabric8ConfigMapPropertySourceLocator` with the `@Retryable` annotations and return those beans only when `spring.cloud.kubernetes.config.retry.enabled=true` and `spring.cloud.kubernetes.config.fail-fast=true`?

**Comment by wind57** [comment]
First of all, we already have `@Retryable` on the `ConfigMapPropertySourceLocator/SecretsPropertySourceLocator` which is extended by `Fabric8ConfigMapPropertySourceLocator`, on par with you so far, I guess.

Second, we need those beans irrespective of `spring.cloud.kubernetes.config.retry.enabled` and `spring.cloud.kubernetes.config.fail-fast`, simply because even without these flags, we still need to query kubeapi-server for those secrets and configmaps. 

To be fair, I kind of don't follow you here a bit, may be I miss something. 

My idea in the PR is this:

- by default, you will get a property source implementation that will : a) not retry b) not fail fast.
- if you enable `fail-fast`, source implementation will : a) fail as soon as an exception is encountered b) will not retry in case of that exception
- if you enable `fail-fast` and `retry`, source implementation will: a) fail as soon as an exception is encountered b) will retry in case of that exception
- you could enable retry and disable fail-fast: it would make no sense and will actually not do anything, but you could nevertheless.

If I miss something in your thought process, do let me know please. thx for your input.

**Comment by ryanjbaxter** [comment]
Is the ultimate problem here that retry is being enabled by https://github.com/spring-cloud/spring-cloud-consul/blob/main/spring-cloud-consul-core/src/main/java/org/springframework/cloud/consul/ConsulAutoConfiguration.java#L109?

This in turn causes the `@Retryable` annotation here https://github.com/spring-cloud/spring-cloud-kubernetes/blob/main/spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/ConfigMapPropertySourceLocator.java#L88 
to blow up because the configuration in https://github.com/spring-cloud/spring-cloud-kubernetes/blob/main/spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/KubernetesBootstrapConfiguration.java#L49 is not run due to `spring.cloud.kubernetes.config.fail-fast` being false?

**Comment by wind57** [comment]
> Is the ultimate problem here that retry is being enabled by https://github.com/spring-cloud/spring-cloud-consul/blob/main/spring-cloud-consul-core/src/main/java/org/springframework/cloud/consul/ConsulAutoConfiguration.java#L109?

Yes. 

> This in turn causes the @Retryable annotation here https://github.com/spring-cloud/spring-cloud-kubernetes/blob/main/spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/ConfigMapPropertySourceLocator.java#L88
to blow up because the configuration in https://github.com/spring-cloud/spring-cloud-kubernetes/blob/main/spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/KubernetesBootstrapConfiguration.java#L49 is not run due to spring.cloud.kubernetes.config.fail-fast being false?

Yes. 

---

Enabling `spring.cloud.kubernetes.config.fail-fast=true` is not correct either, _just_ so that retryable would work. Does my proposal now make sense?

**Comment by ryanjbaxter** [comment]
I'm not saying your solution won't work, I am just wondering if there is an easier solution.

What If we remove the `@Retryable` annotations from the existing `ConfigMapPropertySourceLocator` classes and create specific retryable versions of the classes that we then create beans for if `fail-fast=true` and `retry=true`?  I feel like that *could be* just a bit of additional configuration.  Just a thought.

**Comment by wind57** [comment]
let me see if I understood you correctly. 

 - we will drop `@Retryable` in `ConfigMapPropertySourceLocator`. 
 - we will create a new class : `Fabric8ConfigMapRetryablePropertySourceLocator`. It will have:

```
@Retryable(interceptor = "kubernetesConfigRetryInterceptor")
protected MapPropertySource getMapPropertySource ...
```
- we will keep `Fabric8ConfigMapPropertySourceLocator`
- `Fabric8ConfigMapRetryablePropertySourceLocator` vs `Fabric8ConfigMapPropertySourceLocator` will be chosen based on the flags that you mentioned.

am I right?

**Comment by wind57** [comment]
if you look into `ConfigMapPropertySourceLocator::locate`, it will call `getMapPropertySourceForSingleConfigMap` that it turn will call `getMapPropertySource` - that we are trying to retry. Since this is a direct method call, no proxy is going to be created. So unfortunately, as much as I like the plan - it will not work. 

We can probably work around that, but creating proper beans in place - but the changes are not going to be small at all. :(
